From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: El_Frod0 <m.maljarcik02@gmail.com>
Date: Sat, 3 May 2025 15:12:50 +0200
Subject: [PATCH] Fix PlayerReadyArrowEvent cancellation desync


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 330b21946564e6a7b463a258c02fee3f91e0f057..29091ac15201be3fe359e9397a72930efe2287f4 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -2351,13 +2351,16 @@ public abstract class Player extends LivingEntity {
     }
 
     // Paper start - PlayerReadyArrowEvent
-    protected boolean tryReadyArrow(ItemStack bow, ItemStack itemstack) {
-        return !(this instanceof ServerPlayer) ||
-                new com.destroystokyo.paper.event.player.PlayerReadyArrowEvent(
-                    ((ServerPlayer) this).getBukkitEntity(),
-                    org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(bow),
-                    org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)
-                ).callEvent();
+    // We pass a result mutable boolean in to allow the caller of this method to know if the event was cancelled.
+    protected boolean tryReadyArrow(ItemStack bow, ItemStack itemstack, final org.apache.commons.lang3.mutable.MutableBoolean cancelled) {
+        if (!(this instanceof final ServerPlayer serverPlayer)) return true;
+        final boolean notCancelled = new com.destroystokyo.paper.event.player.PlayerReadyArrowEvent(
+            serverPlayer.getBukkitEntity(),
+            org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(bow),
+            org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)
+        ).callEvent();
+        if (!notCancelled) cancelled.setValue(true);
+        return notCancelled;
     }
     // Paper end - PlayerReadyArrowEvent
 
@@ -2366,13 +2369,14 @@ public abstract class Player extends LivingEntity {
         if (!(stack.getItem() instanceof ProjectileWeaponItem)) {
             return ItemStack.EMPTY;
         } else {
-            Predicate<ItemStack> predicate = ((ProjectileWeaponItem) stack.getItem()).getSupportedHeldProjectiles().and(item -> tryReadyArrow(stack, item)); // Paper - PlayerReadyArrowEvent
+            final org.apache.commons.lang3.mutable.MutableBoolean anyEventCancelled = new org.apache.commons.lang3.mutable.MutableBoolean(); // Paper - PlayerReadyArrowEvent
+            Predicate<ItemStack> predicate = ((ProjectileWeaponItem) stack.getItem()).getSupportedHeldProjectiles().and(item -> tryReadyArrow(stack, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
             ItemStack itemstack1 = ProjectileWeaponItem.getHeldProjectile(this, predicate);
 
             if (!itemstack1.isEmpty()) {
                 return itemstack1;
             } else {
-                predicate = ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles().and(item -> tryReadyArrow(stack, item)); // Paper - PlayerReadyArrowEvent
+                predicate = ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles().and(item -> tryReadyArrow(stack, item, anyEventCancelled)); // Paper - PlayerReadyArrowEvent
 
                 for (int i = 0; i < this.inventory.getContainerSize(); ++i) {
                     ItemStack itemstack2 = this.inventory.getItem(i);
@@ -2382,6 +2386,7 @@ public abstract class Player extends LivingEntity {
                     }
                 }
 
+                if (anyEventCancelled.booleanValue() && !this.abilities.instabuild && this instanceof final ServerPlayer player) this.resyncUsingItem(player); // Paper - resync if no item matched the Predicate
                 return this.abilities.instabuild ? new ItemStack(Items.ARROW) : ItemStack.EMPTY;
             }
         }
